src/lib/storage/buffer/storage_region.cpp

```diff
@@ -0,0 +1,166 @@
+#elif __linux__
+  const int flags = O_RDWR | O_CREAT | O_DIRECT | O_DSYNC;
+#endif
+  const int fd = open(file_name.string().c_str(), flags, 0666);
+  if (fd < 0) {
+    const auto error = errno;
+    close(fd);
+    Fail("Error while opening backing file at " + file_name.string().c_str() + ": " + strerror(error) +
+         ". Did you open a file on tmpfs or a network mount?");
+  }
+
+// In comparison to O_DIRECT on Linux, we need to use fcntl(F_NOCACHE) for direct IO. See comment above.
+#ifdef __APPLE__
+  if (fcntl(fd, F_NOCACHE, 1) == -1) {
+    const auto error = errno;
+    close(fd);
+    Fail("Error while setting F_NOCACHE on __APPLE__: " + strerror(error));
+  }
+#endif
+
+  return fd;
+}
+
+void StorageRegion::write_page(PageID page_id, std::byte* data) {
+  Assert(page_id.valid(), "PageID must be valid");
+  const auto handle = _file_handles[static_cast<uint64_t>(page_id.size_type())];
+  const auto byte_count = page_id.byte_count();
+  const auto offset = handle.offset + byte_count * page_id.index();
+  DebugAssertPageAlignment(data);
+  if (pwrite(handle.fd, data, byte_count, offset) < 0) {
+    const auto error = errno;
+    Fail("Error while writing to StorageRegion: " + strerror(error));
+  }
+  _total_bytes_written.fetch_add(byte_count, std::memory_order_relaxed);
+}
+
+void StorageRegion::read_page(PageID page_id, std::byte* data) {
+  Assert(page_id.valid(), "PageID must be valid");
+  const auto handle = _file_handles[static_cast<uint64_t>(page_id.size_type())];
+  const auto byte_count = page_id.byte_count();
+  const auto offset = handle.offset + byte_count * page_id.index();
+  DebugAssertPageAlignment(data);
+  if (pread(handle.fd, data, byte_count, offset) < 0) {
+    const auto error = errno;
+    Fail("Error while reading from StorageRegion: " + strerror(error));
+  }
+  _total_bytes_read.fetch_add(byte_count, std::memory_order_relaxed);
+}
+
+size_t StorageRegion::memory_consumption() const {
+  return sizeof(*this);
+}
+
+std::array<StorageRegion::FileHandle, PAGE_SIZE_TYPES_COUNT> StorageRegion::_open_file_handles_in_directory(
+    const std::filesystem::path& path) {
+  DebugAssert(std::filesystem::is_directory(path), "StorageRegion path must be a directory");
+  auto array = std::array<StorageRegion::FileHandle, PAGE_SIZE_TYPES_COUNT>{};
+
+  const auto now = std::chrono::system_clock::now();
+  const auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
+
+  for (auto page_size_type_index = size_t{0}; page_size_type_index < PAGE_SIZE_TYPES_COUNT; ++page_size_type_index) {
+    const auto file_name = path / ("hyrise-buffer-pool-" + std::to_string(timestamp) + "-type-" +
+                                   std::to_string(page_size_type_index) + ".bin");
+    array[page_size_type_index] = {_open_file_descriptor(file_name), file_name};
+    // We initialize the file with 32 MiB.
+    std::filesystem::resize_file(file_name, 1UL << 25);
```

Comment:
When/how is this ever resized? Let's say I run SF 1000.

Reply:
So we always use something like 300 MB for the StorageRegion?

Reply:
> When/how is this ever resized? Let's say I run SF 1000.

if you use pwrite to write data to a position beyond the current end of the file, the OS will extend the file size to include the new data.